plugins/metalink/utils.c

```diff
@@ -720,128 +721,172 @@ TDNFParseUrlTag(
-                if(!xmlContValue)
-                {
-                    dwError = ERROR_TDNF_METALINK_PARSER_MISSING_FILE_SIZE;
-                    pr_err("XML Parser Error:File size is missing: %s", size);
-                    BAIL_ON_TDNF_ERROR(dwError);
-                }
-                size = (char*)xmlContValue;
-                if(sscanf(size, "%ld", &(ml_ctx->size)) != 1)
-                {
-                    dwError = ERROR_TDNF_INVALID_PARAMETER;
-                    pr_err("XML Parser Warning: size is invalid value: %s\n", size);
-                    BAIL_ON_TDNF_ERROR(dwError);
-                }
-            }
-            else if(!strcmp((const char*)node->name, TAG_NAME_HASH))
-            {
-                dwError = TDNFParseHashTag(ml_ctx, node);
-                BAIL_ON_TDNF_ERROR(dwError);
-            }
-            else if(!strcmp((const char*)node->name, TAG_NAME_URL))
-            {
-                dwError = TDNFParseUrlTag(ml_ctx, node);
-                BAIL_ON_TDNF_ERROR(dwError);
-            }
+            elementInfo->dwError = ERROR_TDNF_METALINK_PARSER_MISSING_FILE_SIZE;
+            pr_err("XML Parser Error:File size is missing: %s", size);
+            BAIL_ON_TDNF_ERROR(elementInfo->dwError);
         }
-        if (node->children) {
-            TDNFXmlParseData(ml_ctx, node->children, filename);
+        if(sscanf(size, "%ld", &(elementInfo->ml_ctx->size)) != 1)
+        {
+            elementInfo->dwError = ERROR_TDNF_INVALID_PARAMETER;
+            pr_err("XML Parser Warning: size is invalid value: %s\n", size);
+            BAIL_ON_TDNF_ERROR(elementInfo->dwError);
         }
-        node = node->next;
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_HASH))
+    {
+        elementInfo->dwError = TDNFParseHashTag(userData, val, len);
+        BAIL_ON_TDNF_ERROR(elementInfo->dwError);
+    }
+    else if(!strcmp(elementInfo->startElement, TAG_NAME_URL) && len > MIN_URL_LENGTH)
+    {
+        elementInfo->dwError = TDNFParseUrlTag(userData, val, len);
+        BAIL_ON_TDNF_ERROR(elementInfo->dwError);
     }
 
 cleanup:
-    if(xmlContValue)
-    {
-        xmlFree(xmlContValue);
-        xmlContValue = NULL;
+    if(size != NULL){
+       TDNF_SAFE_FREE_MEMORY(size);
     }
-    return dwError;
+    return;
 error:
     goto cleanup;
 }
 
+void
+TDNFXmlParseEndElement(
+    void *userData,
+    const char *name
+    )
+{
+    struct MetalinkElementInfo* elementInfo = (struct MetalinkElementInfo*)userData;
+    elementInfo->endElement = name;
+
+    if(elementInfo->dwError != 0)
+       BAIL_ON_TDNF_ERROR(elementInfo->dwError);
+
+cleanup:
+     return;
+error:
+     goto cleanup;
+}
+
 uint32_t
 TDNFMetalinkParseFile(
     TDNF_ML_CTX *ml_ctx,
-    int fd,
+    FILE *file,
     const char *filename
     )
 {
     uint32_t dwError = 0;
-    xmlDoc *doc = NULL;
-    xmlNode *root_element = NULL;
-
-    if(!ml_ctx || (fd <= 0) || IsNullOrEmptyString(filename))
+    struct stat st;
```

Comment:
Use `struct stat st = {0}` - this would initialize all elements to 0.

Reply:
Done